cpp/src/arrow/util/variant.h

```diff
@@ -17,17 +17,363 @@
+    return {};
+  }
+};
+
+}  // namespace detail
+
+template <typename... T>
+class Variant : detail::variant_storage<T...>,
+                detail::conditional_t<
+                    detail::all<(std::is_copy_constructible<T>::value &&
+                                 std::is_copy_assignable<T>::value)...>::value,
+                    detail::explicit_copy_constructor,
+                    detail::delete_copy_constructor>::template type<Variant<T...>>,
+                detail::member_constructor<Variant<T...>, 0, T...> {
+  static_assert(detail::all<(std::is_nothrow_move_constructible<T>::value &&
+                             std::is_nothrow_move_assignable<T>::value)...>::value,
+                "valueless_by_exception is not supported");
+
+  template <typename U>
+  static constexpr uint8_t index_of() {
+    return detail::member_constructor<Variant<T...>, 0, T...>::index_of(
+        detail::type_constant<U>{});
+  }
+
+ public:
+  using default_type = typename util::detail::first<T...>::type;
+  static_assert(std::is_nothrow_default_constructible<default_type>::value,
+                "valueless_by_exception, non-default constructible are not supported");
+
+  Variant() noexcept { construct_default(); }
+
+  Variant(const Variant& other) = default;
+  Variant& operator=(const Variant& other) = default;
+
+  using detail::member_constructor<Variant<T...>, 0, T...>::member_constructor;
+  using detail::member_constructor<Variant<T...>, 0, T...>::operator=;
+
+  Variant(Variant&& other) noexcept { move(&other, /*assignment=*/false); }
+
+  Variant& operator=(Variant&& other) noexcept {
+    move(&other, /*assignment=*/true);
+    return *this;
+  }
+
+  ~Variant() {
+    static_assert(offsetof(Variant, data_) == 0, "(void*)&Variant::data_ == (void*)this");
+    this->destroy();
+  }
+
+  uint8_t index() const noexcept { return this->index_; }
+
+  template <typename U, uint8_t I = index_of<U>()>
+  const U* get() const noexcept {
+    return index() == I ? reinterpret_cast<const U*>(this) : NULLPTR;
+  }
+
+  template <typename U, uint8_t I = index_of<U>()>
+  U* get() noexcept {
+    return index() == I ? reinterpret_cast<U*>(this) : NULLPTR;
+  }
+
+  template <typename U, typename... A, uint8_t I = index_of<U>()>
+  void emplace(A&&... args) try {
+    this->destroy();
+    new (this) U(std::forward<A>(args)...);
+    this->index_ = I;
+  } catch (...) {
+    construct_default();
+    throw;
+  }
+
+  void swap(Variant& other) noexcept {  // NOLINT google-runtime-references
+    Variant tmp = std::move(other);
+    other = std::move(*this);
+    *this = std::move(tmp);
+  }
+
+ private:
+  void move(Variant* other, bool assignment) noexcept {
+    using impl_t = void(void*, void*);
+    static std::array<impl_t*, sizeof...(T)> impl = {detail::move_construct<T>...};
+
+    if (assignment) {
+      this->destroy();
+    }
+
+    auto index = other->index_;
+    impl[index](this, other);
```

Comment:
Are you sure this kind of function table access will be inlined by the compiler as a switch/case?

Reply:
Clang 10 succeeds: https://godbolt.org/z/ab8Gaf

Moreover, function tables are also the approach taken by vendored/variant.hpp for c++11 when last I checked

Reply:
I wouldn't call that succeed. `print_using_table` uses an indirect jump while `print_using_switch` does inline comparisons.

Reply:
You could do something like:
```c++
template <typename V, uint8_t I, typename H, typename... T>
struct member_constructor<V, I, H, T...> : member_constructor<V, I + 1, T...> {
// ...
void move_to(uint8_t index, Variant* target) {
  if (index == I) {
    detail::move_construct(target, this);
  } else {
    member_constructor<V, I + 1, T...>::move_to(index, target);
  }
}
```

```

Reply:
https://godbolt.org/z/jM6Tzx

Reply:
Will do, thanks